python/pyarrow/tests/test_udf.py

```diff
@@ -0,0 +1,458 @@
+
+
+def test_non_uniform_input_udfs():
+
+    def unary_scalar_function(ctx, scalar1, array1, scalar2):
+        coeff = pc.call_function("add", [scalar1, scalar2])
+        return pc.call_function("multiply", [coeff, array1])
+
+    in_types = {"scalar1": pc.InputType.scalar(pa.int64()),
+                "scalar2": pc.InputType.array(pa.int64()),
+                "scalar3": pc.InputType.scalar(pa.int64()),
+                }
+    func_doc = {
+        "summary": "multi type udf",
+        "description": "desc"
+    }
+    pc.register_scalar_function(unary_scalar_function,
+                                "multi_type_udf", func_doc,
+                                in_types,
+                                pa.int64())
+
+    res = pc.call_function("multi_type_udf",
+                           [pa.scalar(10), pa.array([1, 2, 3]), pa.scalar(20)])
+    assert res == pa.array([30, 60, 90])
+
+
+def test_nullary_functions():
+
+    def gen_random(ctx):
+        import random
+        val = random.randint(0, 10)
+        return pa.scalar(val)
+
+    func_doc = {
+        "summary": "random function",
+        "description": "generates a random value"
+    }
+
+    pc.register_scalar_function(gen_random, "random_func", func_doc,
+                                {},
+                                pa.int64())
+
+    res = pc.call_function("random_func", [])
+    assert res.as_py() >= 0 and res.as_py() <= 10
+
+
+def test_output_datatype():
+    def add_one(ctx, array):
+        ar = pc.call_function("add", [array, 1])
+        ar = ar.cast(pa.int32())
+        return ar
+
+    func_name = "py_add_to_scalar"
+    in_types = {"array": pc.InputType.array(pa.int64())}
+    out_type = pa.int64()
+    doc = {
+        "summary": "add function scalar",
+        "description": "add function"
+    }
+    pc.register_scalar_function(add_one, func_name, doc,
+                                in_types, out_type)
+
+    func = pc.get_function(func_name)
+
+    assert func.name == func_name
+
+    expected_expr = "Expected output type, int64," \
+        + " but function returned type int32"
+
+    with pytest.raises(pa.lib.ArrowTypeError, match=expected_expr):
+        pc.call_function(func_name, [pa.array([20, 30])])
+
+
+def test_output_value():
+    def add_one(ctx, array):
+        ar = pc.call_function("add", [array, 1])
+        ar = ar.cast(pa.int32())
+        return ar
+
+    func_name = "test_output_value"
+    in_types = {"array": pc.InputType.array(pa.int64())}
+    out_type = {}
+    doc = {
+        "summary": "test output value",
+        "description": "test output"
+    }
+
+    expected_expr = "DataType expected, got <class 'dict'>"
+
+    with pytest.raises(TypeError, match=expected_expr):
+        pc.register_scalar_function(add_one, func_name, doc,
+                                    in_types, out_type)
+
+
+def test_output_type():
+    def add_one(ctx, array):
+        return 42
+
+    func_name = "add_to_scalar_as_py"
```

Comment:
What's the convention for naming test functions here? Can you try to use a consistent convention?